### ISSUE TITLE
Publish Koreader for Android

### ISSUE THREAD

**Issue by Hzj-jie** [koreader/koreader#2040]
- KOReader version:
- Device:
#### Issue

Is there any reason we do not publish Koreader in Google play store? If it's simply because nobody has a Google developer account, I could help.

**Comment by Frenzie** [comment]
Afaik mostly because it's alpha quality.

**Comment by houqp** [comment]
Yep, I don't think we are ready for android play yet. There are two major blockers:

* we have many bug repots tagged with android that need to be resolved
* a properly home screen which @chrox is working on

And I think it's better to start with f-droid first, get some feedback, then move to the play store. Otherwise, we might ended up getting many bad reviews :P

**Comment by Frenzie** [comment]
I prefer F-Droid anyway.

**Comment by houqp** [comment]
Even for publishing in f-droid, we still need to clear the first blocker :)

**Comment by gerroon** [comment]
It would be nice to also have more frequent Android builds, either through F-droid or github releases

**Comment by nico202** [comment]
Any news? Especially for the f-droid build

**Comment by Frenzie** [comment]
Unfortunately nothing's really changed. No one has enough time to work on the Android port and it still doesn't work on many devices.

**Comment by nico202** [comment]
Ok thanks

**Comment by dimqua** [comment]
https://gitlab.com/fdroid/rfp/issues/205

**Comment by gerroon** [comment]
I think that koreader is ready for the initial F-droid release. I have been testing it on multiple devices (lollipop, nougat, kitkat) and so far so good.

**Comment by Frenzie** [comment]
@dimqua Anything specific missing?

**Comment by taosxx** [comment]
Is there actually anything in KOReader that doesn't comply with [F-Droids inclusion criteria](https://f-droid.org/docs/Inclusion_Policy/)?

**Comment by Frenzie** [comment]
Not that I know of. Some of the artwork is http://creativecommons.org/licenses/by/3.0 but that shouldn't be an issue. If there are any concrete actionable points I'll try to address them, of course, but I don't want to directly take on the time investment or responsibility right now.

**Comment by IzzySoft** [comment]
As far as I see, there are [some building issues](https://gitlab.com/fdroid/rfp/issues/205#note_76481497). Did you meanwhile update to a newer SDK level? Recent ones no longer support the `android` command.

**Comment by Frenzie** [comment]
@IzzySoft Where specifically? If you use `kodev build android` then for convenience it'll automatically set up NDK r15c if you don't have the `$NDK` or `$ANDROID_NDK` environment variables set (yes, the one with the `android` command, see [here](https://github.com/koreader/koreader-base/pull/510#discussion_r139290793)). This is otherwise unrelated to the build process. I would expect any Docker image or VM dedicated to building Android software to have those in place already?

**Comment by Frenzie** [comment]
@IzzySoft My bad, I was mistaken.

https://github.com/koreader/android-luajit-launcher/blob/66af56c35d32b9c5bd96e97732dd80c1490da481/Makefile#L27

@pazos Do you want to fix that up or shall I? I won't be able to do it today.

**Comment by Frenzie** [comment]
@IzzySoft Actually, what is one supposed to replace this with since it seems to go beyond the usual idiotic random changes? Gradle?

**Comment by IzzySoft** [comment]
@Frenzie Gradle should be good (most apps we build these days are using that). Not being a developer myself, @rudloff is the one who can better answer that (he's also the one who encountered the error).

**Comment by Rudloff** [comment]
Yes, Gradle is the recommended way to build Andorid apps nowadays.

**Comment by pazos** [comment]
@IzzySoft: currently we are limited to SDK platform tools 25.3 since it is the last one with support for ant. You can install it everywhere with the sdk manager. I doubt that fdroid doesn't have an ant-aware build script in their environment.

**Comment by pazos** [comment]
the makefile can be reworked (replace "android" command with the "sdkmanager") but reworking the application to be built with gradle is a pain in the ass :sob:. Even that we have a "minimal gradle wrapper" it just import task for ant, so the limitation on sdk still applies.

**Comment by Rudloff** [comment]
I think our build server still has the old android/ant platform tools. (But I don't have them locally.)

We usually ask app developers to update to Gradle if they can, because it is then easier for us to maintain the builds.
But if it is too much work for you, I will have a look at how we can build this.

**Comment by pazos** [comment]
@Rudloff 

> We usually ask app developers to update to Gradle if they can, because it is then easier for us to maintain the builds.

It should take a while since nobody here is an android expert or want to spend time on something that just works, you know if is not broken don't fix it :nerd_face: 

But, anyways, most of KOReader is built via CMake, like in other platforms, and the java code is just a thin wrapper to be able to use methods not exposed to native activities. The possible benefits of switching to gradle is just being able to build with newer android sdk platform tools.

> But if it is too much work for you, I will have a look at how we can build this.

Building right now is as simple as install prerequisites (including ant) and call `./kodev release android` for generating an armeabi-v7a apk, or call `ANDROID_ARCH=x86 ./kodev release android` for generating a intel apk.

The build system can be called even if there is no sdk/ndk installed on the system, since it will fetch one and install/use it. If you have a supported sdk/ndk already installed you can pass them to ./kodev.

I'm building with `NDK=/opt/android-ndk SDK=/opt/android-sdk ./kodev release android`

If you want to give it a try you are more than welcome, by default we built debug apks only, but this can be switched. I guess you would want to build properly signed non-debuggable apks. 

@Frenzie: Other thing that we probably need to disable is OTA updates, since Fdroid will handle the (signed) updates by itself.

**Comment by Rudloff** [comment]
The app builds correctly locally, but when I try to build it in our build server VM, it does not find some headers (that should be provided by the NDK):
```
In file included from host/buildvm_asm.c:6:0:
host/buildvm.h:9:23: fatal error: sys/types.h: No such file or directory
 #include <sys/types.h>
```
Do you know what could cause this?
(The full logs are [here](https://gist.github.com/Rudloff/280098e4e79759b5d9068d464aef45ba) if you want to have a look.)

**Comment by Frenzie** [comment]
> Building right now is as simple as install prerequisites

It's ever so slightly harder than just that due to having to set up the Transifex client, assuming you want translations. Also see #3754, which I haven't really had much time to look into.

@Rudloff LuaJIT bootstraps itself with Lua. It's all pretty complicated but the tl;dr is that on Debubuntu you need `libc6-dev-i386` and/or `linux-libc-dev:i386` for the host.

**Comment by Rudloff** [comment]
Thanks for the explanation, I added `gcc-multilib` to the build recipe and now it builds correctly.

The only big issue remaining is that we don't accept debuggable APKs:
```
ERROR: Could not build app org.koreader.launcher: APK is debuggable
```
As you guessed, we would need to build an unsigned release APK.

As for the translations, I suggest we focus on publishing an English build first and figure out how we can add translations later.

**Comment by pazos** [comment]
@Rudloff:

> ERROR: Could not build app org.koreader.launcher: APK is debuggable,

If building with ant is allowed in Fdroid we might want to remove android:debuggable="true". Ant will enable this flag on debug builds and disable it on release builds.

not sure about android:largeHeap, which I think is for old devices, but we need android:vmSafeMode="true" to disable Dalvik JIT / Art AOT.

This should work for the moment:
```diff
--- a/platform/android/luajit-launcher/AndroidManifest.xml
+++ b/platform/android/luajit-launcher/AndroidManifest.xml
@@ -13,7 +13,6 @@
     <uses-permission android:name="android.permission.CHANGE_WIFI_STATE"/>
     <uses-permission android:name="android.permission.WAKE_LOCK"/>
     <application android:icon="@drawable/icon" android:label="@string/app_name"
-        android:hasCode="true" android:debuggable="true"
         android:largeHeap="true" android:vmSafeMode="true">
         <meta-data android:name="android.max_aspect" android:value="2.1" />
         <activity android:name="org.koreader.launcher.MainActivity"
diff --git a/Makefile b/Makefile
index 0a24f02..3f52769 100644
--- a/platform/android/luajit-launcher/Makefile
+++ b/platform/android/luajit-launcher/Makefile
@@ -20,7 +20,7 @@ apk: local.properties project.properties
        git submodule update
        ./mk-luajit.sh $(ANDROID_FULL_ARCH)
        ndk-build ANDROID_FULL_ARCH=$(ANDROID_FULL_ARCH)
-       ant debug
+       ant release
         #gradle debug
```

**Comment by pazos** [comment]
Please wait a little bit. 

@Frenzie: I don't know much about the build system. Can we make that `kodev release android`  builds a release apk and use `kodev build android` to generate a debug one. ?

probably we want to add a couple of targets to luajit-launcher Makefile (release and debug instead of apk) and use `make debug` for nightlies/circle.

**Comment by Frenzie** [comment]
@Rudloff I'm taking a quick look right now. I'm afraid that's not all there is to it, but if you want to get a working build today you should probably apply your own patch since we'll have to fix it up a bit more. The basics as posted by @pazos work, for our purposes something more along the following lines but for yours the earlier patch will do just as well:

```diff
diff --git a/AndroidManifest.xml b/AndroidManifest.xml
index d354891..be229bb 100644
--- a/AndroidManifest.xml
+++ b/AndroidManifest.xml
@@ -13,7 +13,6 @@
     <uses-permission android:name="android.permission.CHANGE_WIFI_STATE"/>
     <uses-permission android:name="android.permission.WAKE_LOCK"/>
     <application android:icon="@drawable/icon" android:label="@string/app_name"
-        android:hasCode="true" android:debuggable="true"
         android:largeHeap="true" android:vmSafeMode="true">
         <meta-data android:name="android.max_aspect" android:value="2.1" />
         <activity android:name="org.koreader.launcher.MainActivity"
diff --git a/Makefile b/Makefile
index 0a24f02..cb77282 100644
--- a/Makefile
+++ b/Makefile
@@ -20,8 +20,12 @@ apk: local.properties project.properties
        git submodule update
        ./mk-luajit.sh $(ANDROID_FULL_ARCH)
        ndk-build ANDROID_FULL_ARCH=$(ANDROID_FULL_ARCH)
+ifeq ($(DEBUG), 1)
        ant debug
-        #gradle debug
+       #gradle debug
+else
+       ant release
+endif
 
 local.properties project.properties:
        android update project --path . -t android-$(SDKAPI_MIN)
```

If you want to patch it yourself, there's also this line:

https://github.com/koreader/koreader/blob/b7eb2f7e7fb8d64b76306e42d80a041eeb285068/Makefile#L368-L370

The luajit-launcher patch combined with something like this in the main repo should do the trick:

```diff
diff --git a/Makefile b/Makefile
index 5ba5ff4b..a717c712 100644
--- a/Makefile
+++ b/Makefile
@@ -365,8 +365,8 @@ androidupdate: all
        cd $(INSTALL_DIR)/koreader && 7z a -l -mx=1 \
                ../../$(ANDROID_LAUNCHER_DIR)/assets/module/koreader-$(VERSION).7z * \
                -x!resources/fonts -x!resources/icons/src -x!spec
-       $(MAKE) -C $(ANDROID_LAUNCHER_DIR) apk
-       cp $(ANDROID_LAUNCHER_DIR)/bin/NativeActivity-debug.apk \
+       $(MAKE) -C $(ANDROID_LAUNCHER_DIR) apk DEBUG=$(KODEBUG)
+       cp $(ANDROID_LAUNCHER_DIR)/bin/NativeActivity-release-signed.apk \
                koreader-android-$(MACHINE)-$(VERSION).apk
 
 debianupdate: all
```

Please let us know if there's anything else. Btw, hopefully @pazos or I will work the whole thing out this weekend, but I think that should be all you need to get a working package right now. :-)

@pazos
There are two more alternative filenames there, `NativeActivity-release-unsigned.apk` and  `NativeActivity-release-signed.apk`. The former is actively developer hostile, since unlike a debug apk it won't install for testing. It might be okay for our own release signed with jarsigner (and possibly for something similar like F-Droid) but that's about it. Even for our own release I consider it belligerent because you can also get the unsigned builds from GitLab if you want. So this will require some thought to deal with properly, also so as not to pick up a different apk than the one that was just generated.

**Comment by Frenzie** [comment]
@pazos 

> probably we want to add a couple of targets to luajit-launcher Makefile (release and debug instead of apk) and use make debug for nightlies/circle.

I was thinking something like this might be cleaner?

```diff
diff --git a/Makefile b/Makefile
index 0a24f02..d1b6d46 100644
--- a/Makefile
+++ b/Makefile
@@ -14,14 +14,18 @@ ANDROID_FULL_ARCH?=armeabi-v7a
 # Minimum SDK API is 19, required to use View.VERSION_CODES.KITKAT
 SDKAPI_MIN=$(shell [ ${NDKABI} -ge 19 ] && echo -n ${NDKABI} || echo -n 19)
 
-apk: local.properties project.properties
+apk-base: local.properties project.properties
        git submodule init
        git submodule sync
        git submodule update
        ./mk-luajit.sh $(ANDROID_FULL_ARCH)
        ndk-build ANDROID_FULL_ARCH=$(ANDROID_FULL_ARCH)
+
+debug: apk-base
        ant debug
-        #gradle debug
+
+release: apk-base
+       ant release
 
 local.properties project.properties:
        android update project --path . -t android-$(SDKAPI_MIN)
```

**Comment by Rudloff** [comment]
We are not in a hurry and I would prefer not having to maintain a patch on our side.
So if you plan to work on this in the near future, we can certainly wait :slightly_smiling_face: 

> There are two more alternative filenames there, NativeActivity-release-unsigned.apk and NativeActivity-release-signed.apk.

We will definitely need the unsigned one. Our build process signs the APK separately.

**Comment by Frenzie** [comment]
@pazos 
> I think we want to ship debug apks as nightlies (without debug symbols in KO) and let Fdroid (or particular users) to release properly signed apks.

We could ship debug builds instead of the current nightlies if we wanted, but it'd make them slower, larger, and most important different from the actual release, which sounds like a bad idea to me.

NB Debug or not has nothing to do with signing it. We'll sign them, period.

**Comment by pazos** [comment]
debug builds are signed automatically. AFAIK It is not possible to build a debug apk without signing. But yeah, your last message makes sense.

Anyways, people can't switch between Fdroid and our releases without uninstalling the application, since the keys don't match.

**Comment by Frenzie** [comment]
> Anyways, people can't switch between Fdroid and our releases without uninstalling the application, since the keys don't match.

Of course, but that's exactly why all of our own builds are and will be signed properly. :-)

I do think losing the ability to just install a GitLab artifact (after uninstalling any existing copies) would be a loss, but I guess I should just accept that loss… because Android.

**Comment by IzzySoft** [comment]
> Anyways, people can't switch between Fdroid and our releases without uninstalling the application, since the keys don't match.

Unless you go for "reproducible builds". But that requires additional efforts.

**Comment by KenMaltby** [comment]
As I remember it, just making changes inside the .apk KOReader folder or the res folder, made it necessary to re-sign the .apk.  I don't see how that would change any.  (It's been awhile since I bothered to make any changes to the .apk download before installing it on my phone.)

**Comment by Frenzie** [comment]
If we were to set the tar/zip/7z/etc. modification time to the Git commit time it just might be deterministic already, give or take the order in which files are added to the archive. We don't currently care about either of those, after all. Perhaps that's too optimistic, but except for that very last step of the process I can't immediately think of any other reason they would not be binary identical.

Conversely, like @KenMaltby said, I'm not quite following what that has to do with F-Droid? I'd certainly be willing to at the very least implement some preliminary steps toward reproducible builds by ensuring the order in which files are added to the archive and setting the timestamp to the git commit time. 

I say Git commit time because reproducible builds are nice for security and all, but I still dislike downloading an archive that came out today with timestamps that say something like 1980, or even something more recent like 2015. ;-)

**Comment by Rudloff** [comment]
> I'm not quite following what that has to do with F-Droid?

See [here](https://f-droid.org/en/docs/Reproducible_Builds/). Our build system can compare the APK we build with the upstream APK and if they match we publish the upstream (signed) APK.
This is the only way we can provide APKs that have upstream signature.

However, it is entirely optional and usually requires a lot of work because Android build tools don't always provide a deterministic output.
@IzzySoft was only suggesting that to answer the different signature issue that was raised.
(But honestly, F-Droid users are used to not be able to replace the APK with one from upstream.)

**Comment by Frenzie** [comment]
Thanks for the docs! If I understand correctly that would effectively require duplicating the F-Droid tc, or just straight-up using an F-Droid Docker image. That might not necessarily be as much work as it sounds like. Except for using Debian Stretch instead of Ubuntu 16.04, in principle [this](https://gitlab.com/fdroid/ci-images-base/blob/921296dea754da06b7ed27b46bd0827aa0c50110/Dockerfile) is fairly similar to our own collection of Docker images. As you might imagine our order is somewhat reversed (i.e., the base image doesn't set up anything for Android) and for regular everyday builds translations could be a bit of a sticking point. Anyway, reproducible builds in that sense doesn't seem worth spending much time on right now. :-)

> A standard part of the Android build process is to run some kind of PNG optimization tool, like aapt singleCrunch or pngcrush. These do not provide deterministic output, it is still an open question as to why.

I wasn't expecting that, btw. @pazos we can probably just disable this regardless; I think I already processed the PNGs with optipng.

**Comment by Frenzie** [comment]
@pazos I don't think there's anything else here that needs to be taken care of?

The `koreader-*-arm-*.apk` and `koreader-*-i686-*.apk` artifacts are now unsigned "release" APKs.

**Comment by Rudloff** [comment]
Indeed, the issue here is that we can't have two releases with the same version code.

**Comment by pazos** [comment]
@Rudloff: not sure if it matters for checking new updates but the android project is a git submodule of the main repo. Main repo is https://github.com/koreader/koreader and luajit-launcher project is https://github.com/koreader/android-luajit-launcher. After git submodule init & git submodule update the place to check current version number/code is path_to_koreader/platform/android/luajit-launcher.

> Indeed, the issue here is that we can't have two releases with the same version code.

Not sure how to solve that without manually updating version code each month/release.

**Comment by Frenzie** [comment]
We already have both a `versionCode` and a `versionName` of a type that I believe should be good for Android.

versionCode:
https://github.com/koreader/koreader/blob/27b6c1546a620cb651b7bbd9a8d1b242480ddec7/spec/unit/version_spec.lua#L47

versionName:
https://github.com/koreader/koreader/blob/27b6c1546a620cb651b7bbd9a8d1b242480ddec7/spec/unit/version_spec.lua#L29

The Version module doesn't depend on anything other than Lua.

NB I'm saying that based on https://github.com/koreader/android-luajit-launcher/blob/8825b5d9d1c9cc11770c4c8acc099a1b60a16022/AndroidManifest.xml#L4-L5 which would suggest we can just force-feed the right value (whether through Ant or by modifying that file dynamically). I haven't yet looked it up in the Android docs.

**Comment by Frenzie** [comment]
Reading through it a little closer…

> Warning: The greatest value Google Play allows for versionCode is 2100000000.

Well, gosh darn it. But we could use `git rev-list --count HEAD` to enact a very similar trick. On the plus side, it'd probably be simpler to implement that way.

**Comment by Frenzie** [comment]
This versionCode/versionName thing is actually something I'd been wanting to tackle for a while.

Anyway, please let us know if there's anything else.

**Comment by pazos** [comment]
Yep, we just need to disable OTAupdates, which are not working on Android, anyways. But this can wait until @Rudloff 's feedback

**Comment by Rudloff** [comment]
I tried to build 778c42312d66b6548cb5b5dd1284ec6eac95eee1 with our tools.
I get a release APK with a `v2019.02-7-g778c4231_2019-02-04` version name but the version code is `5`.
I think `ANDROID_VERSION` is not set correctly and it uses [this fallback](https://github.com/koreader/android-luajit-launcher/blob/9a597959319c88fa0e5996085dfc64d88d8d9a46/Makefile#L36).

I tried to run `git rev-list --count HEAD` in our build VM and it returns `6445` as expected ; so I am not sure where the problem comes from.

**Comment by Frenzie** [comment]
@Rudloff Sorry, I forgot to ping you. That issue should be fixed.

**Comment by Rudloff** [comment]
I tried with b2740808469801482e2d83e640e89c44aa1f1fb9 and confirm it now works correctly.
Feel free to ping me when you publish a new stable release.

**Comment by pazos** [comment]
@Rudloff @Frenzie : a question.

We want to provide our own updates from nightly builds. That is fine for us but won't match with fdroid desired behaviour of having all aplications verified. So I was thinking about a way to differenciate between Fdroid builds and KOReader builds at build time, so if the package was built for Fdroid we disable the "checking updates" option.

Even if we finally get reproducible builds the same issue still applies. FDroid needs to verify that specified build is exactly the same you get building from source and this verification needs to be done for each new version available.

**Comment by Frenzie** [comment]
Indeed, we should set this to `no` in certain conditions. The only question is what those conditions should be. A simple environment variable at build time (coupled with a flag in kodev) should probably suffice but it'd be interesting to hear what other apps do.

https://github.com/koreader/koreader/blob/6246d44068e80d9a125c412c74bf0ab306ade708/frontend/device/android/device.lua#L24

**Comment by Rudloff** [comment]
The standard way of doing this for Android apps is [build variants](https://developer.android.com/studio/build/build-variants), but an environment variable would be totally OK for us.

**Comment by pazos** [comment]
After reading a bit of build variantes I would go for the enviroment variable, pleeeeease 😭

**Comment by Frenzie** [comment]
It's not completely clear to me how the app would actually respond to these build flavors. Is it done using the `versionNameSuffix`?

In which case, it should probably be something like `ifdef FDROID NAME+=.fdroid endif` on the current line 27.
https://github.com/koreader/android-luajit-launcher/blob/0f0fbe5803399741a11e651512b2e737939cccf9/Makefile#L25-L27

Or probably generically, an ANDROID_NAME_SUFFIX. Anyway, that's what I'm thinking atm.

**Comment by Rudloff** [comment]
`versionNameSuffix` is not required and we actually don't recommend it because it breaks our auto-update system.

Builds flavors work with Gradle building [flavor-specific classes](https://developer.android.com/studio/build/build-variants#sourceset-build). You can have Java code in `src/main/` that is used in every flavor and code `src/[flavor]/` that is specific to a flavor. This way the app can behave differently depending on which flavor is built.
(You can see an example [here](https://github.com/fossasia/susi_android/tree/development/app/src), the `YoutubeVid` class is different in the `fdroid` and `playStore` flavors.)

But anyway, they require using Gradle. Since your are using a Makefile for building, it totally makes sense to use an environment variable and use it to make whatever changes are needed.

**Comment by pazos** [comment]
```
ANDROID_ARCH=x86 ANDROID_FLAVOUR=fdroid ./kodev release android
```
??

**Comment by Frenzie** [comment]
The principle sounds good, but I don't think flavor needs to mean third party.

**Comment by pazos** [comment]
Ok, so always write variant to a text file and  `hasOTAUpdates = return getFlavour() ~= fdroid end` ?

**Comment by Frenzie** [comment]
Sure, that'll do the trick for now. (Btw, please keep it "flavor" for the code. ;-) )

**Comment by pazos** [comment]
I think that's all in our end. Not sure if @Rudloff implemented the .transifexrc to fetch translations in the build server.

In any case fdroid can now build its own flavor of KOReader for the following android abis:

armeabi-v7a: `ANDROID_FLAVOR=fdroid ./kodev release android`
x86: `ANDROID_ARCH=x86 ANDROID_FLAVOR=fdroid ./kodev release android`

**Comment by Rudloff** [comment]
I didn't work on the translations yet, I would prefer to focus on building and publishing a stable release first (even if it is in English only).

I tried to build v2019.02-154-g5e01a0c2_2019-03-03 with `ANDROID_FLAVOR=fdroid` and it seems to build correctly.
Feel free to ping me when you tag your next release so we can work on publishing it.

**Comment by Frenzie** [comment]
@Rudloff I tagged `v2019.03.1`.

**Comment by Frenzie** [comment]
And thanks to @Rudloff it's in! :-)

<img src=https://user-images.githubusercontent.com/202757/54536196-86f78f00-4990-11e9-8581-40ed97b6725a.png width=40%> <img src=https://user-images.githubusercontent.com/202757/54536198-86f78f00-4990-11e9-8839-beb0fb7822ba.png width=40%>

**Comment by pazos** [comment]
Awesome!!, I dislike the idea of releasing for Google Play or the Amazon Store, so I'm in for closing this issue (or leave it open to track fdroid updates and/or issues).

**Comment by NiLuJe** [comment]
Speaking of, I assume FDroid's thingy is tracking new tags automagically, right?

**Comment by Frenzie** [comment]
Looks like the version is coded in the metadata, but maybe it'll be updated automagically?
https://gitlab.com/fdroid/fdroiddata/merge_requests/4608
https://gitlab.com/fdroid/fdroiddata/merge_requests/4608/diffs

If any F-Droid issues should crop up (and of course there's translation to figure out) I think new specific issues would be clearer.

**Comment by Rudloff** [comment]
Unfortunately, auto update is disabled for this app, because our update system can't handle custom build systems like yours (see [here](https://f-droid.org/fr/docs/Build_Metadata_Reference/#UpdateCheckMode)).
This means someone will have to edit the metadata in order to build a new release.

**Comment by pazos** [comment]
I think `UpdateCheckMode` as `HTTP` can work.

> HTTP - HTTP requests are used to determine the current version code and version name. This is controlled by the UpdateCheckData field, which is of the form urlcode|excode|urlver|exver.

But after a quick search I cannot find any application on the official repo using it. Any hint @Rudloff?

**Comment by Rudloff** [comment]
`HTTP` requires a file that contains the hardcoded version name and version code. But yours are dynamic: https://github.com/koreader/koreader/blob/2f65fb94eca02bbde5200eef36251cca2840857e/kodev#L7
So I don't think we can use it.

**Comment by NiLuJe** [comment]
That could probably be arranged, @Frenzie did something generally similar
to handle otas from withing the app on android/appimage.

The file currently points to a link, IIRC, but another one with just the
stable version should be trivial to do.

**Comment by Frenzie** [comment]
I can easily set up something similar to http://ota.koreader.rocks/koreader-android-latest-stable if desired.

**Comment by Rudloff** [comment]
Oh right, something like this with one line containing the version name and one line containing the version code would probably work.

**Comment by Frenzie** [comment]
@Rudloff So basically, if I add something along these lines to the Makefile:

```diff
diff --git a/Makefile b/Makefile
index 704b3d6d..1d4845e0 100644
--- a/Makefile
+++ b/Makefile
@@ -67,6 +67,9 @@ all: $(if $(ANDROID),,$(KOR_BASE)/$(OUTPUT_DIR)/luajit)
        $(MAKE) -C $(KOR_BASE)
        install -d $(INSTALL_DIR)/koreader
        rm -f $(INSTALL_DIR)/koreader/git-rev; echo "$(VERSION)" > $(INSTALL_DIR)/koreader/git-rev
+ifdef ANDROID
+       rm -f android-fdroid-version; echo -e "$(ANDROID_NAME)\n$(ANDROID_VERSION)" > koreader-android-fdroid-latest
+endif
 ifneq ($(or $(EMULATE_READER),$(WIN32)),)
        cp -f $(KOR_BASE)/ev_replay.py $(INSTALL_DIR)/koreader/
        @echo "[*] create symlink instead of copying files in development mode"
```

Then we'll get the following resulting file `koreader-android-fdroid-latest`:

```
v2019.03-40-g2f65fb94_2019-03-20
6661
```

Which I could designate as a built artifact and automatically put up at `http://ota.koreader.rocks/koreader-android-fdroid-latest`.

Is that about the gist of it? :-)

**Comment by Rudloff** [comment]
Yes, that should do the trick. But you have to only do that for stable releases (our build system will then try to find a Git tag that corresponds to the version name).

**Comment by Frenzie** [comment]
Of course. Note the distinction between http://ota.koreader.rocks/koreader-android-latest-stable and http://ota.koreader.rocks/koreader-android-latest-nightly ;-)

**Comment by Frenzie** [comment]
@Rudloff I tagged `v2019.04` and http://ota.koreader.rocks/koreader-android-fdroid-latest has been automatically populated.

**Comment by Rudloff** [comment]
I think [we require a HTTPS URL](https://gitlab.com/fdroid/fdroidserver/blob/67a8f03fcbfb63ed4f3c9fcb0d9fd859650af5f0/fdroidserver/checkupdates.py#L53) and ota.koreader.rocks seems to be HTTP-only.

**Comment by Frenzie** [comment]
Pinging @houqp

**Comment by houqp** [comment]
I tried putting ota.koreader.rocks behind cloudflare's https proxy, but it was causing OTA upgrade issue (possibly something to do with zsync), so I rolled it back.

@Frenzie I added a new dns record that routes to CF's https proxy: https://fdroid.koreader.rocks. Can you upgrade the nginx server config to serve the same ota directory for this virtual server?

**Comment by NiLuJe** [comment]
Oh, yeah, zsync doesn't support HTTPS.

**Comment by Frenzie** [comment]
@houqp Will do! (Later or tomorrow.)

**Comment by Frenzie** [comment]
@houqp On fdroid.koreader.rocks both http and https now work without issue, but that seems rather redundant when ota.koreader.rocks could do the same thing if you were to set up the CF stuff the same way. ;-)

**Comment by houqp** [comment]
Actually, I forgot whether the zsync issue was due to https or cloudflare
cdn. Could you test quickly to see if fdroid.koreader http works as OTA
server without issue? If so, I can enable cloudflare proxy for ota.koreader
and drop fdroid.

**Comment by NiLuJe** [comment]
If it's served over an OpenStack-ish backend, yeah, that won't work either, even over HTTP.

**Comment by Rudloff** [comment]
I am getting some 403 error:
```
DEBUG: ...requesting https://fdroid.koreader.rocks/koreader-android-fdroid-latest
ERROR: ...checkupdate failed for org.koreader.launcher : HTTP Error 403: Forbidden
```

I think Cloudflare is flagging our user-agent as suspicious...

**Comment by Frenzie** [comment]
@houqp @NiLuJe OTA zsync updating seems to work fine for me using http://fdroid.koreader.rocks

**Comment by NiLuJe** [comment]
@Frenzie : Good to know, thanks!

@Rudloff : No issues over here with either wget/curl/requests from an OVH IP.

**Comment by houqp** [comment]
@Rudloff double checked the access log from cloudflare, the latest blocked request is from 4 days ago:

```
Rule ID	Action Taken	IP Address	Loc.	Host	Date	
IP reputation	Challenge	46.229.168.146	US	koreader.rocks	4 days ago	
IP reputation	Challenge	46.229.168.130	US	build.koreader.rocks	5 days ago	
IP reputation	Challenge	46.229.168.151	US	koreader.rocks	10 days ago
```

Will you be able to get the actual response content from your request? I am curious whether it's flagged by a global security rule that's not controllable from the application firewall setting.

@Frenzie do you want to do more testing or do you want to enable CF proxy for ota.koreader now?

**Comment by Frenzie** [comment]
@houqp I tested it again to successfully update to the latest nightly just now, so I'd say go for it! :-)

(More testing on my end won't show if it'll cause trouble for anybody else, in any case.)

**Comment by houqp** [comment]
Cool, ota.koreader is also behind CF proxy now :)

**Comment by Frenzie** [comment]
@houqp And it works. :-)

**Comment by Rudloff** [comment]
> Will you be able to get the actual response content from your request?

I reproduced the issue with httpie. The request works correctly when my user-agent is `HTTPie/0.9.8`; but when I set it to `Python-urllib/3.5` (what fdroidserver uses), I get a 403 error.

Here is the request:
```http
GET /koreader-android-fdroid-latest HTTP/1.1
Accept: */*
Accept-Encoding: gzip, deflate
Connection: keep-alive
Host: fdroid.koreader.rocks
User-Agent: Python-urllib/3.5
```
And response:
```http
HTTP/1.1 403 Forbidden
CF-RAY: 4c5717872bf26962-CDG
Cache-Control: max-age=10
Connection: keep-alive
Content-Encoding: gzip
Content-Type: text/html; charset=UTF-8
Date: Wed, 10 Apr 2019 19:22:51 GMT
Expect-CT: max-age=604800, report-uri="https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct"
Expires: Wed, 10 Apr 2019 19:23:01 GMT
Server: cloudflare
Set-Cookie: __cfduid=da0175fbde7b19590016160253e951b241554924171; expires=Thu, 09-Apr-20 19:22:51 GMT; path=/; domain=.koreader.rocks; HttpOnly
Transfer-Encoding: chunked
Vary: Accept-Encoding
X-Frame-Options: SAMEORIGIN

<!DOCTYPE html>
<!--[if lt IE 7]> <html class="no-js ie6 oldie" lang="en-US"> <![endif]-->
<!--[if IE 7]>    <html class="no-js ie7 oldie" lang="en-US"> <![endif]-->
<!--[if IE 8]>    <html class="no-js ie8 oldie" lang="en-US"> <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en-US"> <!--<![endif]-->
<head>
<title>Access denied | fdroid.koreader.rocks used Cloudflare to restrict access</title>
<meta charset="UTF-8" />
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1" />
<meta name="robots" content="noindex, nofollow" />
<meta name="viewport" content="width=device-width,initial-scale=1,maximum-scale=1" />
<link rel="stylesheet" id="cf_styles-css" href="/cdn-cgi/styles/cf.errors.css" type="text/css" media="screen,projection" />
<!--[if lt IE 9]><link rel="stylesheet" id='cf_styles-ie-css' href="/cdn-cgi/styles/cf.errors.ie.css" type="text/css" media="screen,projection" /><![endif]-->
<style type="text/css">body{margin:0;padding:0}</style>


<!--[if gte IE 10]><!--><script type="text/javascript" src="/cdn-cgi/scripts/zepto.min.js"></script><!--<![endif]-->
<!--[if gte IE 10]><!--><script type="text/javascript" src="/cdn-cgi/scripts/cf.common.js"></script><!--<![endif]-->



</head>
<body>
  <div id="cf-wrapper">
    <div class="cf-alert cf-alert-error cf-cookie-error" id="cookie-alert" data-translate="enable_cookies">Please enable cookies.</div>
    <div id="cf-error-details" class="cf-error-details-wrapper">
      <div class="cf-wrapper cf-header cf-error-overview">
        <h1>
          <span class="cf-error-type" data-translate="error">Error</span>
          <span class="cf-error-code">1010</span>
          <small class="heading-ray-id">Ray ID: 4c5717872bf26962 &bull; 2019-04-10 19:22:51 UTC</small>
        </h1>
        <h2 class="cf-subheadline">Access denied</h2>
      </div><!-- /.header -->

      <section></section><!-- spacer -->

      <div class="cf-section cf-wrapper">
        <div class="cf-columns two">
          <div class="cf-column">
            <h2 data-translate="what_happened">What happened?</h2>
            <p>The owner of this website (fdroid.koreader.rocks) has banned your access based on your browser's signature (4c5717872bf26962-ua48).</p>
          </div>

          
        </div>
      </div><!-- /.section -->

      <div class="cf-error-footer cf-wrapper">
  <p>
    <span class="cf-footer-item">Cloudflare Ray ID: <strong>4c5717872bf26962</strong></span>
    <span class="cf-footer-separator">&bull;</span>
    <span class="cf-footer-item"><span>Your IP</span>: 2a01:cb0c:3:1a00:d0cb:2ceb:34bc:2a46</span>
    <span class="cf-footer-separator">&bull;</span>
    <span class="cf-footer-item"><span>Performance &amp; security by</span> <a href="https://www.cloudflare.com/5xx-error-landing?utm_source=error_footer" id="brand_link" target="_blank">Cloudflare</a></span>
    
  </p>
</div><!-- /.error-footer -->


    </div><!-- /#cf-error-details -->
  </div><!-- /#cf-wrapper -->

  <script type="text/javascript">
  window._cf_translation = {};
  
  
</script>

</body>
</html>
```

**Comment by houqp** [comment]
@Rudloff good catch! I am able to repro with:

```bash
 curl -v -A "Python-urllib/3.5" https://fdroid.koreader.rocks/koreader-android-fdroid-latest
```

I believe this is from cloudflare's default global firewall rules. Let me see if there is anyway to turn it off.

Unrelated question, have you guys considered setting a custom user-agent for fdroid crawler? I am guessing other people might run into this issue in the future as well.

**Comment by houqp** [comment]
@Rudloff , I have disabled the default Browser Integrity Check for ota.koreader.rocks, this should work without error now:

```
curl -A "Python-urllib/3.5" https://ota.koreader.rocks/koreader-android-fdroid-latest
```

**Comment by Rudloff** [comment]
I enabled auto update in F-Droid: https://gitlab.com/fdroid/fdroiddata/merge_requests/4726

**Comment by pazos** [comment]
Thanks @Rudloff for his work!!  The only thing that prevents me to switch to Fdroid is the lack of translations. Sadly I have no idea about how to fix that other that to create a new repo just for translations. Maybe our lord of the scripts @Frenzie has an idea...

**Comment by Frenzie** [comment]
Not exactly. One of the more obvious solutions might be to return to the situation prior to https://github.com/koreader/koreader/pull/848, except using Weblate instead of Transifex. Because let's face it, I'd rather use Qt Linguist or Poedit than the Transifex web GUI.

You *can* do that of course, but then you have to download it, edit it, upload it… all very bothersome compared to just updating a handful strings in the web GUI now and then, even with the `tx` client. The effort is only potentially worth it if you have at least a couple dozen strings. It's not like Transifex is actively bad or anything, just ever so slightly slower and more cumbersome than I'd like it to be.

But one thing that's tied to our Transifex setup (without an export?) is the translation memory. So if we had a string that was translated in 2014, removed in 2015 and reintroduced in 2019, it's automagically translated. This is platform lock-in, mind you. This would be fairly easy to do manually with a git source search if it were actually part of the git repo. Instead you'd have to extract all the release packages.

Anyway, those are just some tangential ramblings. The most obvious or at least the easiest short-term solution I see is to give a trusted F-Droid Transifex account the necessary rights, if such a thing exists.

**Comment by Rudloff** [comment]
> The most obvious or at least the easiest short-term solution I see is to give a trusted F-Droid Transifex account the necessary rights, if such a thing exists.

Our build recipes are entirely public so that would mean the API key or password would be public as well.

**Comment by NiLuJe** [comment]
Isn't there a "RO" access possible to pull transifex data?

**Comment by Frenzie** [comment]
No, that's the very problem. For some reason Transifex requires privileged access just to pull. But someone should double check if that's still the case in 2019.

**Comment by Frenzie** [comment]
But apparently they came out with some kind of GH integration recently: https://www.transifex.com/blog/2019/github-integration/

**Comment by houqp** [comment]
I just checked transifex, they don't support read-only api token.

GH looks useful for two way sync. It would be nice if we can configure it to also sync partially translated sources.

Another option is we setup a daemon script (or circleci job, github action, etc) that periodically pulls translated sources into a separate repo for public read access.

**Comment by Frenzie** [comment]
Yes, I was thinking the [GitLab Transifex push](https://gitlab.com/koreader/nightly-builds/blob/35b26c8c22cc1b88d81de7c510adabcb51408510/transifex_push.sh) could perhaps take care of something like that with a few extra lines.

**Comment by Frenzie** [comment]
@Rudloff Btw, I added a few screenshots to the README, see https://github.com/koreader/koreader/pull/4959. I don't know the precise format for adding those to F-Droid?

**Comment by Rudloff** [comment]
If you add PNG files in `fastlane/metadata/android/en-US/images/phoneScreenshots/`, F-Droid will use them automatically when publishing the next release (see [here](https://f-droid.org/en/docs/All_About_Descriptions_Graphics_and_Screenshots/#in-the-apps-source-repository)).

**Comment by pazos** [comment]
Hi @Rudloff! We updated the build to use gradle instead of ant. Is still using the same Makefile. Could you test if you're still able to build latest nightly?

```
v2019.06-17-ga335648_2019-07-01
6825
```

Thanks!

**Comment by Rudloff** [comment]
Hello,
We will have to slightly alter our build recipe. I prepared it here: https://gitlab.com/fdroid/fdroiddata/merge_requests/5024

**Comment by pazos** [comment]
A recap of *to-dos* before closing this issue:

- Translations: migration to weblate is in progress (https://github.com/koreader/koreader/issues/3754). The last 2019.11 f-droid version has all the languages. 2019.12 won't because we returned to `tx pull` while weblate instance isn't ready. Should be fixed in 2020.1.

- Archs: F-droid provides a package for arm. They need a different version code for each apk. The version code will be changed in https://github.com/koreader/android-luajit-launcher/pull/203 to add a first number to differenciate between archs. (1) for arm and (2) for x86. Maybe the fdroid script and/or the feed 'koreader-android-fdroid-latest ' need to be updated.

- Metadata: specially screenshots, but a (text) description too: WIP in https://github.com/koreader/koreader/issues/4904. Basically the options are resumed in https://github.com/koreader/koreader/issues/4904#issuecomment-534302086.

**Comment by pazos** [comment]
About releasing on Play Store there are a few issues.

1. Play apps cannot be modified during runtime. So no `patch.lua` or no dictionaries/settings override.
2. App will be required to be sandboxed in November 2020. (enforcing api 29 & scoped storage)
3. The libraries need to follow some requisites: https://android.googlesource.com/platform/bionic/+/master/android-changes-for-ndk-developers.md

They work as they are because the dlopen implementation we have but I'm not sure if Google cares about it or just run readelf agaisnt shared libraries and find they're missing section headers, for example.

I'm not interested at all in a Play Store version of Ko, specially because (1 and 2), but at the same time I'm interested about knowing if the app, as is, would pass the review and would be published.

**Comment by pazos** [comment]
Ok. Translations and metadata are ready. The makefile was updated to generate different version code for differents archs too.

@Rudloff: could you please check if everything is in place?. The file https://gitlab.com/fdroid/fdroiddata/blob/master/metadata/org.koreader.launcher.yml needs to be updated to point to newer translation host: https://hosted.weblate.org/engage/koreader/ and the maintainer note about lack of translations can be removed.

**Comment by Rudloff** [comment]
I tried to build v2019.12, but `git submodule update` fails with the following error:
```
Submodule path 'base': checked out '3813bb1b950abdd75a11b96e237ca9932f3dc1a5'
Submodule path 'base/luafilesystem': checked out '1dfb8c41e8a7e689959baeaf2961437db9615f74'
Submodule path 'base/thirdparty/kpvcrlib/crengine': checked out '4239c45c0916371c090a1f1fc573edbfb2eeba20'
error: Server does not allow request for unadvertised object c848073a4c0c7a08b8a1c29f0686962a6e97af36
Fetched in submodule path 'l10n', but it did not contain c848073a4c0c7a08b8a1c29f0686962a6e97af36. Direct fetching of that commit failed.
```

**Comment by Frenzie** [comment]
I apologize for that. However, the metadata isn't in that tag regardless.

**Comment by Rudloff** [comment]
Oh, then we should wait for the next tag before removing the metadata from the YAML file.

**Comment by pazos** [comment]
> Oh, then we should wait for the next tag before removing the metadata from the YAML file.

@Rudloff: new version ready -> http://ota.koreader.rocks/koreader-android-fdroid-latest

**Comment by Rudloff** [comment]
I removed the description from our metadata: https://gitlab.com/fdroid/fdroiddata/-/merge_requests/6316
This should allow our build tools to fetch it from your repository instead.

**Comment by pazos** [comment]
@Rudloff: thanks! 

Please remove the Maintainer note and replace transifex with weblate when you have time.

**Comment by licaon-kter** [comment]
> Please remove the Maintainer note and replace transifex with weblate when you have time.

Done

But, https://f-droid.org/wiki/page/org.koreader.launcher/lastbuild_7278 ?

**Comment by Frenzie** [comment]
We can make that a little more robust for building older tags.

The fonts are very large, so it's a shallow clone (top commit only).
https://github.com/koreader/koreader/blob/2c102b245be3b3c3ef871ca83851de52936a1826/.gitmodules#L14-L17

So for an older tag you'd have to manually grab more of the submodule if necessary. Our automated builds never need to do that, so it's never been automated.

**Comment by licaon-kter** [comment]
https://ota.koreader.rocks/koreader-android-fdroid-latest
```
v2020.03
7360
```

...but fails since: 
> ERROR: Could not build app org.koreader.launcher: Unexpected version/version code in output; APK: 'v2020.03' / '**17360**',  Expected: 'v2020.03' / '7360'

Where does 17360 come from?

Should I update it to 17360 (and then you need to publish versions higher only) or should I wait for you to fix it?

**Comment by pazos** [comment]
> Where does 17360 come from?

@licaon-kter:  see https://github.com/koreader/koreader/issues/2040#issuecomment-568505517, specially: 

> Archs: F-droid provides a package for arm. They need a different version code for each apk. The version code will be changed in koreader/android-luajit-launcher#203 to add a first number to differenciate between archs. (1) for arm and (2) for x86. Maybe the fdroid script and/or the feed 'koreader-android-fdroid-latest ' need to be updated.

See https://github.com/koreader/android-luajit-launcher/pull/203#issuecomment-567238135 and my next comment.

TL;DR: It is a change done **just** for Fdroid. We don't need it. If you don't plan to release a x86 version we can remove the first number and keep the version matching our number of commits.

**Comment by pazos** [comment]
Pinging @licaon-kter and @Rudloff. ^^

**Comment by Rudloff** [comment]
Unfortunately, our auto update tool does not handle building for multiple architectures.
Would it be possible to return the full ARM version code [here](https://ota.koreader.rocks/koreader-android-fdroid-latest)?

**Comment by pazos** [comment]
> Would it be possible to return the full ARM version code here?

Yeah. That or (partially) revert https://github.com/koreader/android-luajit-launcher/pull/203/files#diff-b67911656ef5d18c4ae36cb6741b7965. Pinging @Frenzie

**Comment by Frenzie** [comment]
So either way there's no x86 version? In that case partially reverting koreader/android-luajit-launcher#203 probably makes more sense to me since in that case it doesn't serve any purpose.

**Comment by pazos** [comment]
@licaon-kter, @Rudloff v2020.04 ready to build -> http://ota.koreader.rocks/koreader-android-fdroid-latest 

Thanks!

**Comment by pazos** [comment]
Fdroid release is updated again, weeeeeeeeeeeeeeeeeee!

![2020-05-01_23-03](https://user-images.githubusercontent.com/975883/80841588-0598a580-8c00-11ea-8a8c-33eca571c342.png)

Thanks again @licaon-kter @Rudloff 

One minor thing: metadata is not picked. Is this because the metadata dir is a submodule? Thanks!

**Comment by Rudloff** [comment]
I'm not sure, our tools should support submodules.

**Comment by pazos** [comment]
> I'm not sure, our tools should support submodules.

Okay!. I'll close the issue but kept notified. Please comment here if you need some change related to metadata or find another issue.

**Comment by sudhirkhanger** [comment]
>App will be required to be sandboxed in November 2020. (enforcing api 29 & scoped storage)

Will KOReader support Scoped Storage?

**Comment by Frenzie** [comment]
It should work if a bit awkwardly (on the backend), see https://github.com/koreader/koreader/pull/5680

**Comment by philippthiele** [comment]
Are there any plans currently to ever publish it on the Play store? What is currently the reason for it not being there? 

I am interested, since on company phones, the company can block any installation if Apps and allow installs from the playstore - which is exactly my situation (this is a very new development of android I think).

This issue seems quite outdated, but didn't want to open a new issue for exactly the same topic.

**Comment by pazos** [comment]
> Are there any plans currently to ever publish it on the Play store? What is currently the reason for it not being there?

No. The app as is cannot be submitted.

It should be possible to submit it if somebody fixes x86_64 build and add support for [app bundles](https://developer.android.com/guide/app-bundle)

Even then GPlay will be qualified to remove our permission over all files rendering the whole app useless.

> I am interested, since on company phones, the company can block any installation if Apps and allow installs from the playstore - which is exactly my situation (this is a very new development of android I think).

People workaround these scenarios with app isolation. You need an app that implements the "work profile" that's available on the PlayStore, like https://play.google.com/store/apps/details?id=com.oasisfeng.island

Then you install apps within the isolated profile.

Please give it a try and tell us if that works for you

**Comment by benoit-pierre** [comment]
> It should be possible to submit it if somebody fixes x86_64 build […]

What's missing? I just did a quick build & test with the android emulator, and it seems to work fine.

**Comment by philippthiele** [comment]
> People workaround these scenarios with app isolation. You need an app that implements the "work profile" that's available on the PlayStore, like https://play.google.com/store/apps/details?id=com.oasisfeng.island
> 
> Then you install apps within the isolated profile.
> 
> Please give it a try and tell us if that works for you

Thanks for suggesting the workaround, I just tried and get an error message roughly translated to English like: It is not possible to add a work profile to this device.

Maybe because there is already a work profile for my work apps existing? Or maybe that is also just blocked by the company policies.

**Comment by pazos** [comment]
> What's missing? I just did a quick build & test with the android emulator, and it seems to work fine.
I can't tell you. Maybe nothing :D 

IIRC it failed a while ago when @NiLuJe added support for arm64: https://github.com/koreader/koreader-base/pull/1540. Probably many of the comments there are outdated since the NDK bump and your magic bits polluting koreader-base :).

**Comment by Frenzie** [comment]
I seem to recall x86_64 working basically okay when I tried it over half a decade ago at the time I added the x86 build; I just didn't see any point in messing about with it beyond that because everything Android x86 was 32-bit anyway.

> What is currently the reason for it not being there?

It would take a lot of effort.

**Comment by poire-z** [comment]
(I thought the reasons were simpler: we don't want to get involved with Google's policies and nightmares, we don't gain any money, so not worth that pain - we don't need more visibility and users and issues and bad reviews.)

**Comment by Frenzie** [comment]
(That's the longer version of what I said. ;-)

**Comment by pazos** [comment]
> Thanks for suggesting the workaround, I just tried and get an error message roughly translated to English like: It is not possible to add a work profile to this device.

> Maybe because there is already a work profile for my work apps existing? Or maybe that is also just blocked by the company policies.

Yeah. Makes sense. Sorry for the false hope :/

> I thought the reasons were simpler: we don't want to get involved with Google's policies and nightmares, we don't gain any money, so not worth that pain we don't need more visibility and users and issues and bad reviews

I agree. We kept it that way because the APK was never in accordance with Google Play policies.

If KOReader mets policies any person could pay 25$ and release a fake "koreader" app. Google won't do shit about it cause there's no trademark.

I vote for the lazy approach, do nothing to release on Play Store :D.

But.

If we got builds that can be published via play console I would say to not to miss the opportunity and launch something. Even a "koreader-demo" that tells users:

Dear user:

This is a limited version. It is provided as is and not supported by any means.

If you want to update to the free, community supported, version go to: url

**Comment by Frenzie** [comment]
That's not going to happen (nor is it relevant to this issue). But there's an app called [Obtainium](https://github.com/ImranR98/Obtainium) which can be used to get updates straight from the source.

**Comment by gd4c** [comment]
@Frenzie Can you share a link to the discussion? I want to know more, please.

**Comment by Frenzie** [comment]
It pretty much speaks for itself: removing patch support from F-Droid takes special effort. See https://github.com/koreader/koreader/issues/6297

**Comment by shuvashish76** [comment]
> It should be possible to submit it if somebody fixes x86_64 build and add support for [app bundles](https://developer.android.com/guide/app-bundle)
>
> Even then GPlay will be qualified to remove our permission over all files rendering the whole app useless.

> (I thought the reasons were simpler: we don't want to get involved with Google's policies and nightmares, we don't gain any money, so not worth that pain - we don't need more visibility and users and issues and bad reviews.)

https://nextcloud.com/blog/nextcloud-android-file-upload-issue-google/
It's nice that KOR preferred not to publish at PlayStore by choice and later [cry-over Google policies](https://www.theregister.com/2025/05/13/nextcloud_play_store_complaint/) just like Nextcloud. Unless KOR would have face exact same issue as Nextcloud.
(Or other issues e.g. auto-removing KDE Connect from user device with _Play Protect_ without user knowledge.)

**Comment by pazos** [comment]
> It's nice that KOR preferred not to publish at PlayStore by choice and later [cry-over Google policies](https://www.theregister.com/2025/05/13/nextcloud_play_store_complaint/) just like Nextcloud. Unless KOR would have face exact same issue as Nextcloud.

We do nothing to achieve play support. That's our first barrier :)
If that barrier falls it is better to have a crippled version from a legit source than a few crappy clones lying around.

**Comment by shuvashish76** [comment]
> Google wants to verify the identity of all developers who distribute apps on Android, **even if it’s outside the Play Store**.

- https://www.androidauthority.com/android-developer-verification-requirements-3590911/
- https://developer.android.com/developer-verification
- https://www.reddit.com/r/fossdroid/comments/1n0171i/google_wants_to_verify_the_identity_of_all/

**Comment by cardboardcpu** [comment]
If Google eventually ends up enforcing a either the facto or the jure ban on "sideloading", could GPlay support be reconsidered?

Incidentally, this could bring support to devices enrolled in the Advance Protection Program.

**Comment by Frenzie** [comment]
> could GPlay support be reconsidered?

It may be worth keeping in mind that I'm utterly indifferent towards it. The result of a reconsideration may not be what you're looking for.

**Comment by shuvashish76** [comment]
https://blog.luanti.org/2026/08/27/luanti-dmca-tracer-ai/
Luanti removed from Google Play due to baseless AI copyright notice
